crates/language-server/src/goto.rs

```diff
@@ -0,0 +1,87 @@
+use fxhash::FxHashMap;
+use hir::{
+    hir_def::{scope_graph::ScopeId, PathId, TopLevelMod, ItemKind},
+    visitor::{prelude::LazyPathSpan, Visitor, VisitorCtxt},
+    HirDb,
+};
+// use hir_analysis::{name_resolution::EarlyResolvedPath, HirAnalysisDb};
+
+use crate::db::LanguageServerDatabase;
+use common::diagnostics::Span;
+use hir::span::LazySpan;
+
+pub(crate) type GotoEnclosingPath = (PathId, ScopeId);
+pub(crate) type GotoPathMap = FxHashMap<Span, GotoEnclosingPath>;
+
+pub struct PathSpanCollector<'db> {
+    // You don't need to collect scope id basically.
+    path_map: GotoPathMap,
+    db: &'db LanguageServerDatabase,
```

Comment:
I tried using dynamic dispatch `db: &'db dyn LanguageServerDb`, but struggled with the compiler later on when I needed to use `db.as_spanned_hir_db()`.  So now I wonder if this change would interfere with salsa somehow?

Reply:
I improve these marker traits usability in https://github.com/ethereum/fe/pull/911; please refer to the comment in the PR. It will resolve this problem.